src/v/s3/client.cc

```diff
@@ -775,4 +767,34 @@ void client_pool::release(ss::shared_ptr<client> leased) {
     _cvar.signal();
 }
 
+void client_pool::load_credentials(cloud_roles::credentials&& credentials) {
+    if (unlikely(!_apply_credentials)) {
+        _apply_credentials = ss::make_lw_shared(
+          cloud_roles::make_credentials_applier(std::move(credentials)));
+        init();
+        // We signal the waiter only after the client pool is initialized, so
+        // that any upload operations waiting are ready to proceed.
+        _credentials_var.signal();
+    } else {
+        _apply_credentials->reset_creds(std::move(credentials));
+    }
+}
+
+ss::future<> client_pool::wait_for_credentials() {
+    while (!_apply_credentials && !_gate.is_closed()
+           && !_as.abort_requested()) {
+        co_await _credentials_var.wait(
+          // In the predicate we make sure that the pool is initialized as well
```

Comment:
there shouldn't be any spurious wakes if you're using `wait` method that accepts a callback, it will loop internally until the predicate is true

Also, probably we need a timeout here. Or maybe we don't need it if it's guaranteed that we will get the credentials eventually or crash the app in a different place.

Reply:
My comment is probably misleading, in the initial code I had the predicate to only wait for credential optional to initialise. 

If we only wait for the credentials  to be initialised then we can have a spurious wake which happens between credential being initialised and pool being populated. 

I added the `_pool.empty()` condition to avoid that. Will update the comment to be clearer as there will be no spurious wake now.

Reply:
fixed comment